### PR TITLE
Sample Code & Animation for Flow Widget

### LINKED ISSUE #10146: User request for an example of Flow widget
Looks like we don't have an example of using Flow.

FYI @HansMuller

### PR DESCRIPTION
**Pull request by Piinks**
## Description

Adds sample code and embeds the animation of the widget in-action for `Flow`.

## Related Issues

Fixes #10146

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
 ///    children.
 ///  * The [catalog of layout widgets](https://flutter.dev/widgets/layout/).
+///
+/// {@tool snippet --template=freeform}
+///
+/// This example uses the [Flow] widget to create a menu that opens and closes
+/// as it is interacted with. The [FlowDelegate] uses the [AnimationController]
+/// to paint the children of the [Flow].
+///
+/// ![A menu created using the Flow widget opens and closes.](https://flutter.github.io/assets-for-api-docs/assets/widgets/flow_menu.mp4)
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: When the menu opens, the house icon has a blue background and the background if the exclamation mark is amber. When the menu closes, the house becomes amber and the exclamation mark blue. It looks a little strange. Is that expected?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
 ///  * [CustomMultiChildLayout], which uses a delegate to position multiple
 ///    children.
 ///  * The [catalog of layout widgets](https://flutter.dev/widgets/layout/).
+///
+/// {@tool snippet --template=freeform}
+///
+/// This example uses the [Flow] widget to create a menu that opens and closes
+/// as it is interacted with. The [FlowDelegate] uses the [AnimationController]
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: What do you mean by "uses the AnimationController to paint..."?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///   }
+/// }
+///
+/// class FlowMenu extends StatefulWidget {
+///   @override
+///   _FlowMenuState createState() => _FlowMenuState();
+/// }
+///
+/// class _FlowMenuState extends State<FlowMenu>
+///   with SingleTickerProviderStateMixin {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: indentation here looks strange. I believe this line needs to be intended by four space to show that it is a continuation of the previous line and not the class body.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///       upperBound: 1000,
+///       duration: const Duration(milliseconds: 250),
+///       vsync: this,
+///     );
+///   }
+///
+///   Widget buildItem(int k, IconData v) {
+///     return GestureDetector(
+///       onTap: () {
+///         if (k != 4) active = k;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: Why can 4 not be active?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 5]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///         menuAnimation.value == 1000
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: 1000? Can you just check if the animation status?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 6]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///           ? menuAnimation.reverse()
+///           : menuAnimation.forward();
+///         setState(() {});
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: Empty setStates are a little smelly. Can you put whatever is modifying the state in here? Or insert a comment describing where the state is changed implicitly that requires this empty setState statement?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 7]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///       },
+///       child: Container(
+///         width: MediaQuery.of(context).size.width / 5,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: What's the 5? https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-mysterious-and-magical-numbers-that-lack-a-clear-derivation

File: packages/flutter/lib/src/widgets/basic.dart [hunk 8]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///     super.initState();
+///     menuAnimation = AnimationController(
+///       lowerBound: 1,
+///       upperBound: 1000,
+///       duration: const Duration(milliseconds: 250),
+///       vsync: this,
+///     );
+///   }
+///
+///   Widget buildItem(int k, IconData v) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: can you give k and v more meaningful names?

Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: Should this be a FlowMenuItem widget?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart]: Oh, i see. The amber icon is the "last selected" icon - too bad you can't see the tapping finger in the video... Maybe add this to the example description above?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 9]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+/// class FlowMenu extends StatefulWidget {
+///   @override
+///   _FlowMenuState createState() => _FlowMenuState();
+/// }
+///
+/// class _FlowMenuState extends State<FlowMenu>
+///   with SingleTickerProviderStateMixin {
+///   AnimationController menuAnimation;
+///   int active = 1;
+///   final Map<int, IconData> menuItems = <IconData>[
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: Why not just a list? you can access every element by the index via `menuItems[index]`

File: packages/flutter/lib/src/widgets/basic.dart [hunk 10]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///     Icons.settings,
+///     Icons.menu,
+///   ].asMap();
+///
+///   @override
+///   void initState() {
+///     super.initState();
+///     menuAnimation = AnimationController(
+///       lowerBound: 1,
+///       upperBound: 1000,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: why 1000?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 11]
```diff
@@ -4706,6 +4706,135 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///
+///   @override
+///   void paintChildren(FlowPaintingContext context) {
+///     double dx = 0.0;
+///     for (int i = 0; i < context.childCount; ++i) {
+///       dx = context.getChildSize(i).width * i;
+///       context.paintChild(
+///         i,
+///         transform: Matrix4.translationValues(
+///           dx * 0.001 * menuAnimation.value,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: Why set the upper bound to 1000 if it gets scaled down to 1 again here?

File: packages/flutter/lib/src/widgets/basic.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/basic.dart]: I can make that clearer. And perhaps add an InkWell to the animation to better illustrate it too. I will revisit and address your other comments as well. Thanks!

File: packages/flutter/lib/src/widgets/basic.dart (hunk 4 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/basic.dart]: The item at the fourth index is the icon to open and close the menu. The active state is for the other button in the menu to indicate interaction. Will make this clearer. 😉

**Review by Piinks — COMMENTED** [review]
Dependent on [updated diagram](https://github.com/flutter/assets-for-api-docs/pull/74) landing.

**Comment by Piinks** [comment]
Diagram has landed.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 12]
```diff
@@ -4706,6 +4706,133 @@ class Wrap extends MultiChildRenderObjectWidget {
...
+///     );
+///   }
+/// }
+///
+/// class FlowMenu extends StatefulWidget {
+///   @override
+///   _FlowMenuState createState() => _FlowMenuState();
+/// }
+///
+/// class _FlowMenuState extends State<FlowMenu>
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart]: nit: I'd probably put this all on one line.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**